File: src/main/java/no/nav/syfo/sts/StsConsumer.java
package no.nav.syfo.sts;

import no.nav.syfo.metric.Metrikk;
import org.springframework.beans.factory.annotation.*;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import static no.nav.syfo.sts.StsToken.setExpirationTime;
import static no.nav.syfo.util.RestUtils.basicCredentials;

@Service
public class StsConsumer {

    private final Metrikk metrikk;
    private final String password;
    private final RestTemplate restTemplate;
    private final String url;
    private final String username;

    private StsToken cachedOidcToken = null;

    @Autowired
    public StsConsumer(
            Metrikk metrikk,
            @Value("${srv.password}") String password,
            @Qualifier("scheduler") RestTemplate restTemplate,
            @Value("${security.token.service.rest.url}") String url,
            @Value("${srv.username}") String username
    ) {
        this.metrikk = metrikk;
        this.password = password;
        this.restTemplate = restTemplate;
        this.url = url;
        this.username = username;
    }

    public String token() {
        if (StsToken.shouldRenew(cachedOidcToken)) {
            metrikk.tellHendelse("call_sts");

            HttpEntity request = new HttpEntity<>(authorizationHeader());

            try {
                ResponseEntity<StsToken> response = restTemplate.exchange(
                        getStsTokenUrl(),
                        HttpMethod.GET,
                        request,
                        StsToken.class
                );
                StsToken token = response.getBody();
                setExpirationTime(token);
                cachedOidcToken = response.getBody();
                metrikk.tellHendelse("call_sts_success");
            } catch (HttpClientErrorException e) {
                metrikk.tellHendelse("call_sts_fail");
                throw e;
            }
        }
        return cachedOidcToken.access_token;
    }

    private String getStsTokenUrl() {
        return url + "/rest/v1/sts/token?grant_type=client_credentials&scope=openid";
    }

    private HttpHeaders authorizationHeader() {
        String credentials = basicCredentials(username, password);
        HttpHeaders headers = new HttpHeaders();
        headers.add(HttpHeaders.AUTHORIZATION, credentials);
        return headers;
    }
}

File: src/main/kotlin/no/nav/syfo/sts/StsConsumer.kt
package no.nav.syfo.sts

import no.nav.syfo.metric.Metrikk
import no.nav.syfo.sts.StsToken
import no.nav.syfo.util.RestUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientResponseException
import org.springframework.web.client.RestTemplate

@Service
class StsConsumer @Autowired constructor(
        private val metrikk: Metrikk,
        @Value("\${srv.password}") private val password: String,
        @Qualifier("scheduler") private val restTemplate: RestTemplate,
        @Value("\${security.token.service.rest.url}") private val url: String,
        @Value("\${srv.username}") private val username: String
) {
    private var cachedOidcToken: StsToken? = null
    fun token(): String {
        if (StsToken.shouldRenew(cachedOidcToken)) {
            val request = HttpEntity<Any>(authorizationHeader())
            try {
                val response = restTemplate.exchange(
                        stsTokenUrl,
                        HttpMethod.GET,
                        request,
                        StsToken::class.java
                )
                cachedOidcToken = response.body
                metrikk.tellHendelse(METRIC_CALL_STS_SUCCESS)
            } catch (e: RestClientResponseException) {
                LOG.error("Request to get STS failed with status: ${e.rawStatusCode} and message: ${e.responseBodyAsString}")
                metrikk.tellHendelse(METRIC_CALL_STS_FAIL)
                throw e
            }
        }
        return cachedOidcToken!!.access_token
    }

    private val stsTokenUrl: String
        get() = "$url/rest/v1/sts/token?grant_type=client_credentials&scope=openid"

    private fun authorizationHeader(): HttpHeaders {
        val credentials = RestUtils.basicCredentials(username, password)
        val headers = HttpHeaders()
        headers.add(HttpHeaders.AUTHORIZATION, credentials)
        return headers
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(StsConsumer::class.java)

        const val METRIC_CALL_STS_SUCCESS = "call_sts_success"
        const val METRIC_CALL_STS_FAIL = "call_sts_fail"
    }
}
